src/aspire/operators/diag_matrix.py

```diff
@@ -0,0 +1,617 @@
+        """
+        Instantiate ones intialized `DiagMatrix` with length `n`.
+        This corresponds to the diag(A) where A is (n,n).
+
+        :param n: Length of diagonal.
+        :param dtype: Datatype, defaults to np.float32.
+        :return: `DiagMatrix` instance.
+        """
+
+        return DiagMatrix(np.ones(n, dtype=dtype))
+
+    @staticmethod
+    def eye(n, dtype=np.float32):
+        """
+        Build a `DiagMatrix` eye (identity) matrix.
+        This is simply an alias for `ones`.
+
+        :param n: Length of diagonal.
+        :param dtype: Datatype, defaults to np.float32.
+        :return: `DiagMatrix` instance.
+        """
+
+        return DiagMatrix.ones(n, dtype=dtype)
+
+    def as_blk_diag(self, partition):
+        """
+        Express `DiagMatrix` as a `BlkDiagMatrix` using `partition`.
+
+        :return: `BlkDiagMatrix`
+        """
+        if self.stack_shape != ():
+            raise RuntimeError(
+                f"as_blk_diag only implemented for singletons at this time, received {self.stack_shape}."
+            )
+
+        B = BlkDiagMatrix(partition, dtype=self.dtype)
+        ind = 0
+        for b, p in enumerate(partition):
+            assert p[0] == p[1]
+            j = p[0]
+            B[b] = np.diag(self._data[ind : ind + j])
+            ind += j
+
+        return B
+
+    # TODO, discuss name.
+    def lr_scale(self, partition, weights=None):
+        """
+        Performs L and R scaling by this `DiagMatrix` instance
+        simultaneously, returning the result as a `BlkDiagMatrix`
+        having `partition`.
+
+        Given `weights` this computes B = w * A * A.T,
+        where A is this `DiagMatrix` instance.
+
+        :partition: Partition of output `BlkDiagMatrix`.
+        :weights: Optional weight vector, defaults to ones.
+        :return: `BlkDiagMatrix`
+        """
+        if self.stack_shape != ():
+            raise RuntimeError(
+                f"lr_scale only implemented for singletons at this time, received {self.stack_shape}."
+            )
+
+        if weights is None:
+            weights = np.ones(self.count, dtype=self.dtype)
+
+        B = BlkDiagMatrix(partition, dtype=self.dtype)
+
+        ind = 0
+        for block, p in enumerate(partition):
+            if p[0] != p[1]:
+                raise RuntimeError(f"Partition block {block} was not square {p}.")
+            j = p[0]
+            Ai = self._data[ind : ind + j].reshape(-1, 1)
+            wi = weights[ind : ind + j]
+            B[block] = wi * Ai * Ai.T
+            ind += j
+
+        if ind != self.count:
+            raise RuntimeError(
+                f"Partition count {ind} does not match DiagMatrix count {self.count}."
+            )
+
+        return B
+
+    def solve(self, b):
+        """
+        For this `DiagMatrix` `a` and diagonal matrix `b`.
+        solve a x = b for `x`.
+
+        :param b: `DiagMatrix`, right hand side.
+        :return: `DiagMatrix`, solution `x`.
```

Comment:
I changed it to always return a DiagMatrix. It could alternatively always return a np array.

Other types of returns would be solving something that is not a DiagMatrix (either a dense or block..).